File: New UI/ProvidersSelection/ProvidersSelectionVC.m
//
//  ProvidersSelectionVC.m
//  linphone
//
//  Created by Norayr Harutyunyan on 3/25/17.
//
//

#import "ProvidersSelectionVC.h"
#import "LoginTVC.h"
#import "Provider.h"

@interface ProvidersSelectionVC () <UITableViewDelegate, UITableViewDataSource> {
    NSMutableArray *providersList;
    Provider *selectedProvider;
}

@property (weak, nonatomic) IBOutlet UIButton *nextButton;
@property (weak, nonatomic) IBOutlet UITableView *tableView;

@end

@implementation ProvidersSelectionVC

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Uncomment the following line to preserve selection between presentations.
    // self.clearsSelectionOnViewWillAppear = NO;
    
    // Uncomment the following line to display an Edit button in the navigation bar for this view controller.
    // self.navigationItem.rightBarButtonItem = self.editButtonItem;

    selectedProvider = nil;
    [self setupUI];
    [self setupDelegateMethods];
    [self setupDataSource];
}

- (void) setupUI {
    [self setupNextButton];
    [self enableNextButton:false];
}

- (void) setupDelegateMethods {
    [self setupTableViewDelegateMethods];
}

- (void) setupTableViewDelegateMethods {
    self.tableView.delegate = self;
    self.tableView.dataSource = self;
}

- (void) setupNextButton {
    self.nextButton.layer.cornerRadius = self.nextButton.frame.size.width / 2.0;
}

- (void) enableNextButton:(BOOL)enable {
    [self.nextButton setEnabled:enable];

    if (enable) {
        self.nextButton.backgroundColor = [UIColor colorWithRed:56.0/255.0 green:197.0/255.0 blue:1.0 alpha:1.0];
    } else {
        self.nextButton.backgroundColor = [UIColor colorWithRed:208.0/255.0 green:211.0/255.0 blue:224.0/255.0 alpha:1.0];
    }
}

- (void) setupDataSource {
    NSURLSessionConfiguration *defaultConfigObject = [NSURLSessionConfiguration defaultSessionConfiguration];
    NSURLSession *defaultSession = [NSURLSession sessionWithConfiguration:defaultConfigObject];
    
    //Create an URLRequest
    NSURL *url = [NSURL URLWithString:@"https://webrtc-provisioning.46labs.com:9443/api/v1/provider/list"];
    NSMutableURLRequest *urlRequest = [NSMutableURLRequest requestWithURL:url];
    
    //Create task
    NSURLSessionDataTask *dataTask = [defaultSession dataTaskWithRequest:urlRequest completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        //Handle your response here
        
        if ((long)((NSHTTPURLResponse*)response).statusCode == 200) {
            NSArray *providers = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingAllowFragments error:nil];
            providersList = [[NSMutableArray alloc] init];
            
            for (NSDictionary *dictProvider in providers) {
                Provider *provider = [[Provider alloc] initWithNSDictionary:dictProvider];
                [providersList addObject:provider];
            }
            
            [self.tableView performSelectorOnMainThread:@selector(reloadData) withObject:nil waitUntilDone:true];
        } else {
            NSLog(@"Problem with geting users");
        }
    }];
    
    [dataTask resume];
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)onButtonNext:(id)sender {
    LoginTVC *loginVC = (LoginTVC*)[UIStoryboard storyboardWithName:@"Login" bundle:nil].instantiateInitialViewController;
    loginVC.provider = selectedProvider;
    [self.navigationController pushViewController:loginVC animated:true];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return providersList.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell" forIndexPath:indexPath];
    
    // Configure the cell...
    Provider *provider = providersList[indexPath.row];
    
    cell.textLabel.text = provider.name;
    cell.textLabel.textColor = [UIColor blackColor];
    
    return cell;
}

- (void) tableView:(UITableView *)tableView didSelectRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    [tableView cellForRowAtIndexPath:indexPath].textLabel.textColor = [UIColor colorWithRed:56.0/255.0 green:197.0/255.0 blue:1.0 alpha:1.0];
    selectedProvider = providersList[indexPath.row];
    [self enableNextButton:true];
}

- (void) tableView:(UITableView *)tableView didDeselectRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    [tableView cellForRowAtIndexPath:indexPath].textLabel.textColor = [UIColor blackColor];
}

/*
// Override to support conditional editing of the table view.
- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}
*/

/*
// Override to support editing the table view.
- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        // Delete the row from the data source
        [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation:UITableViewRowAnimationFade];
    } else if (editingStyle == UITableViewCellEditingStyleInsert) {
        // Create a new instance of the appropriate class, insert it into the array, and add a new row to the table view
    }   
}
*/

/*
// Override to support rearranging the table view.
- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
}
*/

/*
// Override to support conditional rearranging of the table view.
- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the item to be re-orderable.
    return YES;
}
*/

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: New UI/ProvidersSelection/ProvidersSelectionVC.swift
//
//  ProvidersSelectionVC.swift
//  Peeredge
//
//  Created by Karen Muradyan on 3/20/17.
//  Copyright © 2017 Peeredge. All rights reserved.
//

import UIKit

class ProvidersSelectionVC: UIViewController, UITableViewDelegate, UITableViewDataSource {

    @IBOutlet weak var nextButton: UIButton!
    @IBOutlet weak var tableView: UITableView!
    
    fileprivate var providersList = Array<String>()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        setupUI ()
        setupDelegateMethods ()
        setupDataSource ()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
    }
    
    func setupUI () {
        setupNextButton ()
        enableNextButton(enable: false)
    }
    
    func setupDelegateMethods () {
        setupTableViewDelegateMethods ()
    }
    
    func setupTableViewDelegateMethods () {
        tableView.delegate = self
        tableView.dataSource = self
    }
    
    func setupNextButton () {
        nextButton.layer.cornerRadius = nextButton.frame.width / 2
    }
    
    func enableNextButton (enable:Bool) {
        
        if enable {
            nextButton.isEnabled = enable
            nextButton.backgroundColor = UIColor.init(red: 56.0/255.0, green: 197.0/255.0, blue: 250.0/255.0, alpha: 1)
        } else {
            nextButton.isEnabled = enable
            nextButton.backgroundColor = UIColor.init(red: 208.0/255.0, green: 211.0/255.0, blue: 224.0/255.0, alpha: 1)
        }
    }
    
    // MARK: - Setup Data source 
    
    func setupDataSource () {
        // TODO: Make API call to get providers list, providers should be an objects
        providersList = ["Globalgig", "Telmex", "Brasilfone", "BrightLink", "TSI Corp"]
    }

    // MARK: - UITableView delegate methods
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return providersList.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell:UITableViewCell = tableView.dequeueReusableCell(withIdentifier: "cell")! as UITableViewCell
        cell.textLabel?.text = providersList[indexPath.row]
        cell.textLabel?.textColor = UIColor.black
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.cellForRow(at: indexPath)?.textLabel?.textColor = UIColor.init(red: 56.0/255.0, green: 197.0/255.0, blue: 250.0/255.0, alpha: 1)
        enableNextButton(enable: true)
    }
    
    func tableView(_ tableView: UITableView, didDeselectRowAt indexPath: IndexPath) {
        tableView.cellForRow(at: indexPath)?.textLabel?.textColor = UIColor.black
    }

    @IBAction func nextButtonAction(_ sender: Any) {
        let loginVC = UIStoryboard(name: "Login", bundle:nil).instantiateViewController(withIdentifier: "LoginTV") as! LoginTV
        loginVC.providerName = "Telmex"
        self.navigationController?.pushViewController(loginVC, animated: true)
    }
    
}
